filepicker-library/src/io/filepicker/services/ContentService.java

```diff
@@ -252,5 +266,40 @@ private void handleError(RetrofitError error) {
 
         return errorType;
     }
+
+    private static class ProgressTypedFile extends TypedFile {
+
+        private static final int BUFFER_SIZE = 4096;
+
+        interface Listener {
+            void onProgress(float progress);
+        }
+
+        private final Listener mListener;
+
+        public ProgressTypedFile(TypedFile typedFile, Listener listener) {
+            super(typedFile.mimeType(), typedFile.file());
+            mListener = listener;
+        }
+
+        @Override
+        public void writeTo(OutputStream out) throws IOException {
+            byte[] buffer = new byte[BUFFER_SIZE];
+            FileInputStream in = new FileInputStream(file());
+            float written = 0;
+
+            try {
+                int read;
+                while ((read = in.read(buffer)) != -1 && !Thread.interrupted()) {
+                    out.write(buffer, 0, read);
+                    written += read;
+                    mListener.onProgress(written / length());
```

Comment:
I'd suggest to assign the value of `length()` method before the loop, since it won't change inside the loop